src/Examine.Core/ValueSet.cs

```diff
@@ -4,141 +4,317 @@
+            => !Values.TryGetValue(key, out IList<object> values) ? Enumerable.Empty<object>() : values;
+
+        /// <summary>
+        /// Sets the value for the key.
+        /// </summary>
+        /// <param name="key">The key.</param>
+        /// <param name="value">The value.</param>
+        public void SetValue(string key, object value)
+            => Values[key] = new List<object>
+            {
+                value
+            };
+
+        /// <summary>
+        /// Sets the values for the key.
+        /// </summary>
+        /// <param name="key">The key.</param>
+        /// <param name="values">The values.</param>
+        public void SetValues(string key, IEnumerable<object> values)
+            => Values[key] = new List<object>(values);
+
+        /// <summary>
+        /// Adds the value if the key doesn't exist.
+        /// </summary>
+        /// <param name="key">The key.</param>
+        /// <param name="value">The value.</param>
+        /// <returns>
+        ///   <c>true</c> if the value was added; otherwise, <c>false</c>.
+        /// </returns>
+        public bool TryAddValue(string key, object value)
         {
-            return !Values.TryGetValue(key, out var values) ? Enumerable.Empty<object>() : values;
+            if (Values.ContainsKey(key))
+            {
+                return false;
+            }
+
+            Values.Add(key, new List<object>
+            {
+                value
+            });
+
+            return true;
         }
 
         /// <summary>
-        /// Gets a single value for the key
+        /// Adds the values if the key doesn't exist.
         /// </summary>
-        /// <param name="key"></param>
+        /// <param name="key">The key.</param>
+        /// <param name="values">The values.</param>
         /// <returns>
-        /// If there are multiple values, this will return the first
+        ///   <c>true</c> if the values were added; otherwise, <c>false</c>.
         /// </returns>
-        public object GetValue(string key)
+        public bool TryAddValues(string key, IEnumerable<object> values)
         {
-            return !Values.TryGetValue(key, out var values) ? null : values.Count > 0 ? values[0] : null;
+            if (Values.ContainsKey(key))
+            {
+                return false;
+            }
+
+            Values.Add(key, new List<object>(values));
+
+            return true;
         }
 
         /// <summary>
-        /// Helper method to return IEnumerable from a single
+        /// Removes the values for the key.
         /// </summary>
-        /// <param name="i"></param>
-        /// <returns></returns>
-        private static IEnumerable<object> Yield(object i)
+        /// <param name="key">The key.</param>
+        /// <returns>
+        ///   <c>true</c> if the values were removed; otherwise, <c>false</c>.
+        /// </returns>
+        public bool RemoveValues(string key)
```

Comment:
I've added this for consistency with the Get/Set/Add methods, but it might make more sense if there was also an overload that would accept values to remove (instead of all), something like:

```c#
public bool RemoveValues(string key, IEnumerable<object> values)
{
    if (!Values.TryGetValue(key, out IList<object> oldValues))
    {
        return false;
    }

    SetValues(key, oldValues.Except(values));
    return true;
}
```